Added onboarding content from blog post -- missing images for now (#878)
clerk/clerk-docs: docs/guides/add-onboarding-flow.mdx

---
title: How to Add an Onboarding Flow for your Application

> description: Leverage Clerk’s customizable session tokens, publicMetadata and Next’s Middleware to create a robust onboarding experience within a few lines of code.

Author: I think since we're changing this to feel more "tutorial" style these changes make sense 👍

---

# How to Add an Onboarding Flow for your Application

> As part of your onboarding flow, you may want to collect extra information from your user and use it to drive your application state. Let’s walk through a quick example using Next.js and TypeScript to show you how simple implementing an onboarding flow can be.

Author: > "if you are using Clerk's authentication components..."

This doesn't change if you're using the components or Account portal right? Interested in the specific language here which makes me assume as the reader I can only do this with the components flow and custom pages. 

We're doing that in _this_ tutorial, but it's a generally helpful flow either way right?

> As part of your onboarding flow, you may want to collect extra information from your user and use it to drive your application state. Let’s walk through a quick example using Next.js and TypeScript to show you how simple implementing an onboarding flow can be.

Author: This whole change seems to lengthen the description to allow for more of an understanding of the steps we're going to take, but creates a pretty lengthy singular sentence. 

Maybe something like: 

"Onboarding is a crucial part of many authentication flows. Sometimes you need to make sure certain criteria is met and collected before allowing access to parts of your application. With Clerk, you can leverage customizable session tokens, public metadata, and Middleware to properly address these needs in your own app."

> As part of your onboarding flow, you may want to collect extra information from your user and use it to drive your application state. Let’s walk through a quick example using Next.js and TypeScript to show you how simple implementing an onboarding flow can be.

Author: Would you mind just editing in the commit suggestion and I'll commit it @alexisintech 

Looks like only a couple last things before we ship 🚢

> As part of your onboarding flow, you may want to collect extra information from your user and use it to drive your application state. Let’s walk through a quick example using Next.js and TypeScript to show you how simple implementing an onboarding flow can be.

Reviewer: done!

> 

Author: Similar comment to above about specifically mentioning Clerk components

> 

Reviewer: for this paragraph, we are mentioning what this specific guide is going to lead a user to do - so we should actually mention that they will authenticate using AP. I've edited the suggestion!

> 

Reviewer 2: This process is not related to AP or even components. Its applies to any new user whether they signed up or in with AP / components / custom flows.

In this guide, you will learn how to:

1. Add custom claims to your session token
2. Configure your middleware to read session data
3. Update with the user’s onboarding state

To see a working example, check out our sample demonstration app [here](https://sample-onboarding-app.clerkpreview.com/).

<Callout type="info" emoji="💡">
  The examples below have been pared down to the bare minimum to enable you to
  easily customize them to your needs, you can build them with the [Clerk + Next
  Quickstart](https://clerk.com/docs/quickstarts/nextjs) using @clerk/nextjs
  4.29.5 and Next 14.0.4.
</Callout>

> Let’s get started!

Author: Do we want to link to the repo as well? 

I think now that I ask this I wonder...are we going to own a _new_ repo for this? If not, should we just teach to the repo that already exists so we're not overloading with similar content

https://github.com/clerk/clerk-nextjs-onboarding-sample-app

> Let’s get started!

Reviewer: I reached out to you about that link being broken

> Let’s get started!

Author: Fixed the link in the details of the repo, how do we want to proceed here?

> Let’s get started!

Reviewer: updated the callout to include the repository instead of the deployment.

> <img />

Author: All of the screenshots live currently inside of the blog post, you'll see I didn't add them cause I figured you'd say this lol

> <img />

Author: Comfortable with taking the image out if we want to replace with context, let me know if you're adding that or I am

> <img />

Reviewer: we added the context in the introduction: "This guide creates an example onboarding flow that requires users to complete a form before they can access the application. After a user authenticates using Clerk's [Account portal](/docs/account-portal/overview), the user is prompted to fill out a form with an application name and type. Once the user has completed the form, they are redirected to the application dashboard."

there are no necessary screenshots for this guide !

## Add custom claim to your session token

Session tokens are JWTs that are generated by Clerk on behalf of your instance, and contain claims that allow you to store data about a user’s session. With Clerk, when a session token exists for a user, it indicates that the user is authenticated, and the associated claims can be retrieved at any time. [Learn More](https://clerk.com/docs/backend-requests/making/custom-session-token)

> First, navigate to [Sessions](https://dashboard.clerk.com/last-active/sessions?_gl=1*1bx2kkm*_gcl_au*MTcwMzAwNTE2LjE3MDQ5MTA2MjAuMTQ0NTAzMDk3Ny4xNzA1MDM3Njk5LjE3MDUwMzc3MDU.*_ga*OTkyMTU1NTM2LjE2ODk0NzAzMDU.*_ga_1WMF5X234K*MTcxMjE5NDgyMC40OS4xLjE3MTIxOTU0ODQuMC4wLjA.) in your Clerk Dashboard and click the ‘Edit’ button. In the modal that opens, there will be a window where you can augment your session token with custom claims.

Author: Added stepped view for editing the session token

In there, add the following and hit save:

```json
{
  "metadata": "{{user.public_metadata}}"
}
```

<img />

> If you haven’t already, we can make the public metadata type information accessible to our application by adding the following to src/types/globals.d.ts:

Author: To be clear, Next always generates this folder, so if we want to make a callout like this for completeness, do we want to mention this or just assume it doesn't exist always and leave it as is? 

I don't have a strong opinion

> If you haven’t already, we can make the public metadata type information accessible to our application by adding the following to src/types/globals.d.ts:

Reviewer: I don't have this folder generated automatically generated in any of my next apps 😳
where did you see that Next always generates this folder?

> If you haven’t already, we can make the public metadata type information accessible to our application by adding the following to src/types/globals.d.ts:

Reviewer 2: `/types/` is not auto-created, nor is it specifically needed or needed in that location.

```ts
export {};

declare global {
  interface CustomJwtSessionClaims {
    metadata: {
      onboardingComplete?: boolean;
    };
  }
}
```

> We have just added custom data to our session token in the Clerk Dashboard and made those claims accessible to our app. Next, we’ll use clerkMiddleware to redirect the user based on `onboardingComplete` status.

Reviewer: I still don't think this should be here. In a blog maybe, but in docs, this is redundant - the user can simply go to the next step to understand what they're going to do next.

> We have just added custom data to our session token in the Clerk Dashboard and made those claims accessible to our app. Next, we’ll use clerkMiddleware to redirect the user based on `onboardingComplete` status.

Reviewer: I remove this text that's at the end of every step because it's redundant and informal. The user can simply move on to the next step to see what they need to do next. Maybe in blogs this is okay, because blogs are more informal and the writer speaks directly to the user more, but in technical writing, this is not proper prose.

## Configure your Next.js middleware to read session data

Clerk's authMiddleware() allows you ton configure access to your routes with fine grained control. You can also retrieve claims directly from the session and redirect your user accordingly. [Learn More](https://clerk.com/docs/references/nextjs/auth-middleware#usage)

> Add the code sample below to your `src/middleware.ts` file:

Author: Perfect middle ground 🤝

    }

    // User isn't signed in and the route is private -- redirect to sign-in
    if (!userId && !auth.isPublicRoute)
      return redirectToSignIn({ returnBackUrl: req.url });

    // Catch users who doesn't have `onboardingComplete: true` in PublicMetata
    // Redirect them to the /onboading out to complete onboarding
    if (userId && !sessionClaims?.metadata?.onboardingComplete) {
      const onboardingUrl = new URL("/onboarding", req.url);
      return NextResponse.redirect(onboardingUrl);
    }

    // User is logged in and the route is protected - let them view.
    if (userId && !auth.isPublicRoute) return NextResponse.next();

    // If the route is public, anyone can view it.
    if (auth.isPublicRoute) return NextResponse.next();
  },
});

export const config = {
  matcher: ["/((?!.+\\.[\\w]+$|_next).*)", "/", "/(api|trpc)(.*)"],
};
```

Next, create a `layout.tsx` file in `src/app/onboarding` and add the following code to the file. This logic could go in the Middleware, but by adding to the `layout.tsx` to the route the logic remains in one place. This file can also be expanded to handle multiple steps, if multiple steps are required for an onboarding flow.

> ```tsx

Author: add filename to code block

import { auth } from "@clerk/nextjs/server";
import { redirect } from "next/navigation";

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  // Check if a user has completed onboarding
  // If yes, redirect them to /dashboard
  if (auth().sessionClaims?.metadata.onboardingComplete === true) {
    redirect("/dashboard");
  }

  return <>{children}</>;
}
```

> Now that we have the logic for where to direct the user, we’ll need a way to track their onboarding status and note it on their session, let’s dig into that now!

Author: Is this removed in favor to the changes made in "Update publicMetadata based on onboarding state"

> Now that we have the logic for where to direct the user, we’ll need a way to track their onboarding status and note it on their session, let’s dig into that now!

Reviewer: I remove this text that's at the end of every step, because it's redundant. The user can simply move on to the next step to see what they need to do next. Maybe in blogs this is okay, because blogs are more informal and the writer speaks directly to the user more, but in technical writing, this is not proper prose.

## Update publicMetadata based on onboarding state

Updating a user's publicMetadata as they progress through the flow will allow us to recognize when they have successfully completed their onboarding and, per the logic above, are now able to access the application. [Learn More](https://clerk.com/docs/users/metadata)

To do this you need:

- A method in your backend to securely update the user publicMetadata
- A process in your frontend with logic to collect and submit all the information for onboarding. In this guide you’ll use an example form.

## Add userUpdate method to your backend

First, add a method in your backend, that will be called on form submission and update the user’s `publicMetadata` accordingly. The example below uses the `clerkClient` wrapper to interact with the Backend API.

> Under src/app/onboarding/\_actions.ts add the following code snippet:

Author: To be clear, we will need a new section creating the form itself that should come _after_ this addition?

> Under src/app/onboarding/\_actions.ts add the following code snippet:

Reviewer: Nono, we have both the sections we need - we just need to swap their places.
So "Add a form" will go above this one

```tsx
"use server";

import { auth, clerkClient } from "@clerk/nextjs/server";

export const completeOnboarding = async (formData: FormData) => {
  const { userId } = auth();

  if (!userId) {
    return { message: "No Logged In User" };
  }

  try {
    const res = await clerkClient.users.updateUser(userId, {
      publicMetadata: {
        onboardingComplete: true,
        applicationName: formData.get("applicationName"),
        applicationType: formData.get("applicationType"),
      },
    });
    return { message: res.publicMetadata };
  } catch (err) {
    return { error: "There was an error updating the user metadata." };
  }
};
```

> Now that we have a method to securely update our user’s `publicMetadata` we can call this server action from a client side form.

Reviewer: I remove this text that's at the end of every step because it's redundant and informal. The user can simply move on to the next step to see what they need to do next. Maybe in blogs this is okay, because blogs are more informal and the writer speaks directly to the user more, but in technical writing, this is not proper prose.

## Add a form to your frontend

With the backend updateUser method in place, we’ll add a basic page that contains a form to complete the onboarding process.

This example form that will capture an application name (applicationName) and application type of either B2C or B2B (applicationType). This is a very loose example — you can use this step to capture information from the user, sync user data to your database, have the user sign up to a course or subscription, or more.

> To implement this logic, insert the following into your `src/app/onboarding/page.tsx`:

Author: I'm assuming this was referenced here https://github.com/clerk/clerk-docs/pull/878#discussion_r1554179267

We can link up on what's needed here and I'll commit 👍

import * as React from "react";
import { useUser } from "@clerk/nextjs";
import { useRouter } from "next/navigation";
import { completeOnboarding } from "./_actions";

export default function OnboardingComponent() {
  const [error, setError] = React.useState("");
  const { user } = useUser();
  const router = useRouter();

  const handleSubmit = async (formData: FormData) => {
    const res = await completeOnboarding(formData);
    if (res?.message) {
      await user?.reload();
      router.push("/dashboard");
    }
    if (res?.error) {
      setError(res?.error);
    }
  };
  return (
    <div>
      <h1>Welcome</h1>
      <form action={handleSubmit}>
        <div>
          <label>Application Name</label>
          <p>Enter the name of your application.</p>
          <input type="text" name="applicationName" required />
        </div>

        <div>
          <label>Application Type</label>
          <p>Describe the type of your application.</p>
          <input type="text" name="applicationType" required />
        </div>
        {error && <p className="text-red-600">Error: {error}</p>}
        <button type="submit">Submit</button>
      </form>
    </div>
  );
}
```

> ## Wrap Up
> 
> Your onboarding flow is now complete! 🎉 New users who haven’t yet onboarded will now land on your `/onboarding` page and, once they have completed onboarding, will be sent through to the dashboard. By using Clerk, which already handles user authentication, we were able to simplify the process of creating a custom user onboarding flow as well.

Reviewer: ```suggestion
## Wrap up

Your onboarding flow is now complete! 🎉 Users who have not onboarded yet will now land on your `/onboarding` page. New users signing up or signing in to your application will have to complete the onboarding process before they can access your application. By using Clerk, you have streamlined the user authentication and onboarding process, ensuring a smooth and efficient experience for your new users.
```

  // Define which routes are public
  publicRoutes: [],

  afterAuth: async (auth, req: NextRequest) => {
    const { userId, sessionClaims } = auth;

    // For users visiting /onboarding, don't try to redirect
    if (userId && req.nextUrl.pathname === "/onboarding") {
      return NextResponse.next();
    }

    // If the user isn't signed in and the route is private, redirect to sign-in
    if (!userId && !auth.isPublicRoute)
      return redirectToSignIn({ returnBackUrl: req.url });

    // Catch users who do not have `onboardingComplete: true` in their publicMetadata
    // Redirect them to the /onboading route to complete onboarding
    if (userId && !sessionClaims?.metadata?.onboardingComplete) {
      const onboardingUrl = new URL("/onboarding", req.url);
      return NextResponse.redirect(onboardingUrl);
    }

    // If the user is logged in and the route is protected, let them view.
    if (userId && !auth.isPublicRoute) return NextResponse.next();

    // If the route is public, anyone can view it.
    if (auth.isPublicRoute) return NextResponse.next();
  },
});

export const config = {
  matcher: ["/((?!.+\\.[\\w]+$|_next).*)", "/", "/(api|trpc)(.*)"],
};
```

> 

Reviewer: ```suggestion

### Create a layout for the `/onboarding` route

```

```tsx filename="src/app/onboarding/layout.tsx"
import { auth } from "@clerk/nextjs/server";
import { redirect } from "next/navigation";

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  if (auth().sessionClaims?.metadata.onboardingComplete === true) {
    redirect("/dashboard");
  }

  return <>{children}</>;
}
```

## Use `publicMetadata` to track user onboarding state

Each Clerk user has a `User` object that contains a `publicMetadata` property, which can be used to store custom data about the user. This information can be accessed on the client side and can be used to drive application state. [Learn more about public metadata.](/docs/users/metadata#public-metadata)

You can use the user's `publicMetadata` to track the user's onboarding state. To do this, you will create:

- A process in your frontend with logic to collect and submit all the information for onboarding. In this guide, you will create a simple form.
- A method in your backend to securely update the user's `publicMetadata`

> ## Collect user onboarding information

Reviewer: ```suggestion
### Collect user onboarding information
```

1. In your `/onboarding` directory, create an `_actions.ts` file.
2. Add the following code to the file. This file includes a method that will be called on form submission and will update the user's `publicMetadata` accordingly. The following example uses the [`clerkClient`](/docs/references/backend/overview) wrapper to interact with the Backend API and update the user's `publicMetadata`.

```tsx
"use server";

import { auth, clerkClient } from "@clerk/nextjs/server";

export const completeOnboarding = async (formData: FormData) => {
  const { userId } = auth();

  if (!userId) {
    return { message: "No Logged In User" };
  }

  try {
    const res = await clerkClient.users.updateUser(userId, {
      publicMetadata: {
        onboardingComplete: true,
        applicationName: formData.get("applicationName"),
        applicationType: formData.get("applicationType"),
      },
    });
    return { message: res.publicMetadata };
  } catch (err) {
    return { error: "There was an error updating the user metadata." };
  }
};
```

> ## Collect user onboarding information

Reviewer: This section still needs to get swapped with the section above "Update the user's `publicMetadata` in your backend".

This is because we list out what the user needs to do:
> - A process in your frontend with logic to collect and submit all the information for onboarding. In this guide, you will create a simple form.
> - A method in your backend to securely update the user's `publicMetadata`

and creating a form is first. so this section needs to go first
